src/Common/src/System/Security/Cryptography/KeyFormatHelper.cs

```diff
@@ -0,0 +1,620 @@
+            int requiredBytes = encoding.GetByteCount(password);
+            Span<byte> passwordBytes = stackalloc byte[0];
+            byte[] rentedPasswordBytes = Array.Empty<byte>();
+
+            if (requiredBytes > 128)
+            {
+                rentedPasswordBytes = ArrayPool<byte>.Shared.Rent(requiredBytes);
+                passwordBytes = rentedPasswordBytes;
+            }
+            else
+            {
+                passwordBytes = stackalloc byte[requiredBytes];
+            }
+
+            try
+            {
+                fixed (byte* bytePtr = rentedPasswordBytes)
+                {
+                    int written = encoding.GetBytes(password, passwordBytes);
+
+                    if (written != requiredBytes)
+                    {
+                        Debug.Fail("UTF8 encoding length changed between size and convert");
+                        throw new CryptographicException();
+                    }
+
+                    passwordBytes = passwordBytes.Slice(0, written);
+
+                    return WriteEncryptedPkcs8(
+                        password,
+                        passwordBytes,
+                        pkcs8Writer,
+                        pbeParameters);
+                }
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(passwordBytes);
+
+                if (rentedPasswordBytes.Length > 0)
+                {
+                    ArrayPool<byte>.Shared.Return(rentedPasswordBytes);
+                }
+            }
+        }
+
+        internal static AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<byte> passwordBytes,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            return WriteEncryptedPkcs8(
+                ReadOnlySpan<char>.Empty,
+                passwordBytes,
+                pkcs8Writer,
+                pbeParameters);
+        }
+
+        private static unsafe AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<char> password,
+            ReadOnlySpan<byte> passwordBytes,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            ReadOnlySpan<byte> pkcs8Span = pkcs8Writer.EncodeAsSpan();
+
+            PasswordBasedEncryption.InitiateEncryption(
+                pbeParameters,
+                out SymmetricAlgorithm cipher,
+                out string hmacOid,
+                out string encryptionAlgorithmOid,
+                out bool isPkcs12);
+
+            // We need at least one block size beyond the input data size.
+            byte[] encryptedRent = ArrayPool<byte>.Shared.Rent(
+                checked(pkcs8Span.Length + (cipher.BlockSize / 8)));
+
+            Span<byte> encryptedSpan = default;
+            AsnWriter writer = null;
+
+            try
+            {
+                Span<byte> iv = stackalloc byte[cipher.BlockSize / 8];
```

Comment:
This is guaranteed to be relatively small, right?  There's no way this could be some user-written cipher that has a huge block size?

Reply:
Yep.  An enum comes in, only 3DES (64 bit) and AES (128 bit) can come out.  Added an assert.